### ISSUE TITLE
Getting 502 Bad Gateway when backend health is healthy with 200 status code

### ISSUE THREAD

**Issue by sandy-wang** [Azure/application-gateway-kubernetes-ingress#812]
**Describe the bug**
In order to use annotation to set up customer root ca for our application gateway, we upgraded the ingress-azure to the RC 1.2.0-rc1. Then I updated ingress to use the new annotations. Everything on portal is looking good, the backend health shows healthy, but when I browsed my website, I got 502 error, I looked through accesslog, the rows of 502 didn't have much difference from the rows of 200, the only difference is 200 are still using 8080 and http while 502 has 8443 and https.
From the agic pod, it can curl the web page internally using pod's ip.

**To Reproduce**
1. helm upgrade ingress-azure to 1.2.0-rc1
https://github.com/Azure/application-gateway-kubernetes-ingress/releases/tag/1.2.0-rc1
2. Follow the doc https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/docs/annotations.md to update k8s ingress for your application to use appgw-ssl-certificate and appgw-trusted-root-certificate
3. Waited until the http setting and listener got updated, then browser your application web page.

**Ingress Controller details**  (confidential details are masked)
```
< helm list >
ingress-azure           default         6               2020-04-20 20:15:59.3176218 +1000 AEST  deployed        ingress-azure-1.2.0-rc1 1.2.0-rc1

< kubectl get ingress ***-api -o yaml >
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    appgw.ingress.kubernetes.io/appgw-ssl-certificate: ***-tls
    appgw.ingress.kubernetes.io/appgw-trusted-root-certificate: ***-rootca
    appgw.ingress.kubernetes.io/backend-protocol: https
    appgw.ingress.kubernetes.io/use-private-ip: "true"
  labels:
    app: ***-api
  name: ***-api
  namespace: dev
spec:
  rules:
  - host: ***-api-dev.***.***.***
    http:
      paths:
      - backend:
          serviceName: ***-api
          servicePort: 8443
status:
  loadBalancer:
    ingress:
    - ip: ***.***.***.50

<kcg po -n dev -l app=***-api -o wide>
NAME                              READY   STATUS    RESTARTS   AGE   IP              NODE                                NOMINATED NODE   READINESS GATES
***-api-***-w69fg   1/1     Running   0          32m   ***.***.***.53   aks-agentpool-***-vmss000001   <none>           <none>

root@ingress-azure-***-v2fqp:/# curl https://***.***.***.53:8443/actuator/health -k
{"status":"UP"}

<curl https://***-api-dev.***.***.***/>
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>Microsoft-Azure-Application-Gateway/v2</center>
</body>
</html>

{
  "backendAddressPool": {
    "backendAddresses": null,
    "backendIpConfigurations": null,
    "etag": null,
    "id": "/subscriptions/***/resourceGroups/***/providers/Microsoft.Network/applicationGateways/***/backendAddressPools/pool-dev-***-api-8443-bp-8443",
    "name": null,
    "provisioningState": null,
    "resourceGroup": "***",
    "type": null
  },
  "backendHttpSettingsCollection": [
    {
      "backendHttpSettings": {
        "affinityCookieName": null,
        "authenticationCertificates": null,
        "connectionDraining": null,
        "cookieBasedAffinity": null,
        "etag": null,
        "hostName": null,
        "id": "/subscriptions/***/resourceGroups/***/providers/Microsoft.Network/applicationGateways/***/backendHttpSettingsCollection/bp-dev-***-api-8443-8443-***-api",
        "name": null,
        "path": null,
        "pickHostNameFromBackendAddress": null,
        "port": null,
        "probe": null,
        "probeEnabled": null,
        "protocol": null,
        "provisioningState": null,
        "requestTimeout": null,
        "resourceGroup": "***",
        "trustedRootCertificates": null,
        "type": null
      },
      "servers": [
        {
          "address": "***.***.***.53",
          "health": "Healthy",
          "healthProbeLog": "Success. Received 200 status code",
          "ipConfiguration": null
        }
      ]
    }
  ]
},
```

**Comment by sandy-wang** [comment]
I also raised a MS support ticket, and got the root cause located.  Turned out this version of AGIC requires the host name override to be the custom domain name for http setting. And the hostname has to be set and have been set through Ingress rule in AKS. 
https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/docs/annotations.md#backend-hostname

Azure Networking engineer is looking at the potential functional behavior change in agic between 1.0.1-rc3 and 1.2.0-rc1. Because we never had to use backend-hostname when we use custom root ca for http settings,

Closing this ticket.

**Comment by andreaslarfors** [comment]
@sandy-wang Your comment saved me! I was trying to create an App Gateway with E2E TLS and I could not get it to work when not setting the backend hostname. So I set the backend hostname to be the same as the frontend path and it worked!

So this configuration did NOT work:
`---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  labels:
    name: myapp
    component: appgw
  name: myapp-appgw
  namespace: myapp
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/ssl-redirect: "true"
    appgw.ingress.kubernetes.io/backend-protocol: "https"
    appgw.ingress.kubernetes.io/appgw-trusted-root-certificate: "app-gw-backend-cert"
spec:
  tls:
    - secretName: frontend-tls
      hosts:
        - myapp.mydomain.com
  rules:
    - host: myapp.mydomain.com
      http:
        paths:
        - path: /
          backend:
            serviceName: myapp-webserver
            servicePort: 443`

Result: 502 Bad Gateway, but all Health Probes successful.

But adding  appgw.ingress.kubernetes.io/backend-hostname: "myapp.mydomain.com" makes it work, i.e.:
`---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  labels:
    name: myapp
    component: appgw
  name: myapp-appgw
  namespace: myapp
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/ssl-redirect: "true"
    appgw.ingress.kubernetes.io/backend-protocol: "https"
    appgw.ingress.kubernetes.io/backend-hostname: "myapp.mydomain.com"
    appgw.ingress.kubernetes.io/appgw-trusted-root-certificate: "app-gw-backend-cert"
spec:
  tls:
    - secretName: frontend-tls
      hosts:
        - myapp.mydomain.com
  rules:
    - host: myapp.mydomain.com
      http:
        paths:
        - path: /
          backend:
            serviceName: myapp-webserver
            servicePort: 443`

Hope this saves someone else some time!
TL;DR You must specify the backend hostname when using E2E TLS in an App Gateway.
